### PR TITLE
Fix uninitialized variable in x509_crt

### PR DESCRIPTION
**Pull request by agross-oss**
This patch fixes an issue we encountered with more stringent compiler
warnings.  The signature_is_good variable has a possibility of being
used uninitialized.  This patch initializes the variable at the top
of the function.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

We found this issue while using mbedtls inside the Zephyr project.  We recently have started using the x509 features and this exposed this specific issue due to our compiler settings.  I am unsure about backporting as we are using 2.16.0.

**Gatekeeping note:** (added by mpg) this PR should be merge both development and mbedtls-2.16. The issue is not present in mbedtls-2.7 so no need for backports.

### CODE REVIEW

**Comment by hanno-arm** [comment]
Thanks @agross-linaro for your contribution! To my understanding, `signature_is_good` is never used uninitialized, but it's just (too) hard for the compiler to notice that - do you agree? I remember being impressed that some compilers apparently are capable of that amount of reasoning.

**Comment by agross-linaro** [comment]
@hanno-arm well for people using modern compilers on projects which require no warnings these kinds of things pop up.  With all of the continues in that function I can see how this could potentially happen.  But it probably would never happen with the way the function is being used.

I also wonder if tooling like coverity might catch this.

**Comment by hanno-arm** [comment]
@agross-linaro Could you elaborate? Do you see an execution path which would lead to `signature_is_good` being used uninitialized? I don't see one. The only way the branch `parent != NULL` can be taken is if the loop exited through the `break` instruction, which happens after `signature_is_good` is set.

> well for people using modern compilers on projects which require no warnings 

We require no warnings, too, so we'll fix this if some major toolchain warns about it - even if it's unjustified.

File: library/x509_crt.c [hunk 1]
```diff
@@ -2014,7 +2014,8 @@ static int x509_crt_find_parent_in(
 {
     int ret;
     mbedtls_x509_crt *parent, *fallback_parent;
-    int signature_is_good, fallback_signature_is_good;
+    int signature_is_good = 0;
+    int fallback_signature_is_good;
```
Comment (hanno-arm) [inline library/x509_crt.c]: The code might be easier to understand - and avoid unnecessary initialization - if we instead moved the code-block
```
        *r_parent = parent;
        *r_signature_is_good = signature_is_good;
```
to before the `break` in the loop.

**Comment by agross-linaro** [comment]
So this can be bad if:
1) this is a non re-started call to x509_crt_find_parent_in()
2) you have a valid candidate list that has a single entry
3) The x509_crt_check_parent() returns  non zero

With these 3 things, you can use the signature_is_good uninitialized.

**Comment by hanno-arm** [comment]
@agross-linaro No, in this case `parent = parent->next` will set `parent` to `NULL` before leaving the loop.

**Comment by agross-linaro** [comment]
Oh you're right.  Let me keep trying to find the path then =D

**Comment by hanno-arm** [comment]
@agross-linaro Whatever the outcome, the fact that it's not immediately clear is reason enough to change it. See the review suggestion.

File: library/x509_crt.c (hunk 1 above)
Comment (agross-linaro) [reply library/x509_crt.c]: i'm ok with that.  Let me re-test and re-push

**Comment by RonEld** [comment]
CLA valid as part of the Linaro CLA

**Comment by hanno-arm** [comment]
Ping @Patater @sbutcher-arm for gatekeeping.

**Comment by mpg** [comment]
@agross-linaro We usually ask external contributors to add an entry in the ChangeLog crediting themselves. Please add commit to this PR with ChangeLog entry in the Bugfix section, mentioning fixing a compiler warning, with you name (and affiliation) as you want it, plus this issue number.  I'm labeling as "needs: work" as a reminder for that.

Again, thanks for your contribution.

**Comment by agross-linaro** [comment]
Sorry for the delay, I had other pressing concerns.  I added the requested Changelog entry.

**Comment by pfalcon** [comment]
Hello, some updates here: Andy is not with Linaro any longer, that's why I assume this issue got stale. As his colleague, I'd be willing to pick this up and make last-mile changes to make it mergeable. Please let me know if it's ok if I open a separate PR with the changes requested (or if there's some other way). @agross-oss: Please let me know if that sounds ok for you too.

> I would expect Coverity to be able to perform the above reasoning, hence not warn about it.

To elaborate on the nature of the compiler warning described here, compilers perform a static analysis, based on control flow graph (CFG) properties of the code (here, that each *static* use is dominated by a definition). Compilers don't perform analysis of dynamic code properties, like possible actual code execution paths, because it's equivalent of [Halting problem](https://en.wikipedia.org/wiki/Halting_problem).

Bottom line of this theoretical excursion: in software maintenance practice, such warnings can always appear, so thanks mbedTLS team for handling these proactively!

**Comment by pfalcon** [comment]
To avoid possible further slips due to weekend, etc., I proactively posted https://github.com/ARMmbed/mbedtls/pull/2795 . Apologies if that's a bit hasty move, can be closed any time if this PR is going to be revamped.

**Comment by pfalcon** [comment]
This has now been superseded by https://github.com/ARMmbed/mbedtls/pull/2795, https://github.com/ARMmbed/mbedtls/pull/2813 . @mpg, @k-stachowiak, @Patater: Thanks for merging those, and can you please close this PR?

**Comment by mpg** [comment]
Closing for the reasons mentioned by @pfalcon.
